doc/developer/design/20231027_refresh_every_mv.md

```diff
@@ -0,0 +1,164 @@
+
+How do we know how much earlier than the refresh time should we turn on the replica, that is, how much time the refresh will take? In the first version of this feature, we can let the user set this explicitly by something like `REFRESH EVERY <interval> EARLY <interval>`. Later, we should record the times the refreshes take, and infer the time requirement of the next refresh based on earlier ones. Note that this will be complicated by the fact that we have different instance types [that have wildly differing CPU performance](https://materializeinc.slack.com/archives/CM7ATT65S/p1697816482502819).
+
+### Covering the Times Between Creating the MV and the First Refresh
+
+Consider what happens with the materialized view during the time between creating the materialized view and the first refresh. With the solution proposed so far, we would immediately move the upper of the materialized view to the first refresh time. The issue with this is that the materialized view is initially empty, and we would lock in this empty state until the first refresh. This means that querying the materialized view in this time range would always yield an empty result. Note that there was a recent fix for such an issue with sources; we shouldn't introduce now this problem in a new way.
+
+We could solve this as follows: When rounding up times, we would treat times not beyond the original `as_of` of the MV (i.e., the creation time of the MV) specially: we don't round them up, but keep them as it is. (These times should actually be only the `as_of` itself, as sources modify earlier times to the `as_of`.) This would mean that the materialized view's upper would move past the `as_of` only once it has managed to process all input data that happened no later than the creation of the materialized view. If we don't give a replica to the MV initially, then this would simply mean that queries would block on the MV until the first refresh completes. This would surely cause confusion for some users, so to make the user experience more convenient, we could also start up a replica right away when creating the MV, and keep it running until the initial snapshot is complete. This would allow users to examine the contents of the MV soon after creating it, i.e., they wouldn't have to wait until the time of the first refresh.
+
+Note that we currently don't store the original `as_of` of MVs (they get a new `as_of` when starting up a new replica or on envd restarts), so we'd need to store this somewhere for the above to work. I think we'll need to put the `as_of` in the catalog.
```

Comment:
Do we have a similar problem during restarts? If we have some materialized view `mv`, with a non-standard refresh interval and we boot at `boot_ts`, all other materialized views will re-hydrate up to `boot_ts`, but `mv` will not re-hydrate until the next refresh interval. Which means that after a restart, we won't be able to query this `mv` until the next refresh interval. Is that accurate? I'm not actually that familiar with re-hydration.